lib/OpenQA/Shared/Plugin/Gru.pm

```diff
@@ -115,7 +116,66 @@ sub is_task_active ($self, $task) {
 # checks if there are worker registered
 sub has_workers ($self) { !!$self->app->minion->backend->list_workers(0, 1)->{total} }
 
+# For some tasks with the same args we don't need to repeat them if they were
+# enqueued less than a minute ago, like 'git fetch'
+sub _find_existing_gru_task ($self, $task, $args, $jobs) {
+    # warn __PACKAGE__ . ':' . __LINE__ . ": ========================== _find_existing_gru_task($task)\n";
+    my $schema = OpenQA::Schema->singleton;
+    $args = [$args] if ref $args eq 'HASH';
+    my $dtf = $schema->storage->datetime_parser;
+    my $dbh = $schema->storage->dbh;
+    my $sql = q{SELECT id, args, created, state, retries, notes, result FROM minion_jobs
+                WHERE state IN ('inactive', 'active', 'finished')
+                AND created >= ? AND task = ? AND args = ? limit 10};
+    my $sth = $dbh->prepare($sql);
+    my @args = (
+        $dtf->format_datetime(DateTime->now()->subtract(minutes => 1)),
+        'git_clone', OpenQA::Schema::Result::GruTasks->encode_json_to_db($args));
+    $sth->execute(@args);
+    while (my $job = $sth->fetchrow_hashref) {
+        # warn __PACKAGE__ . ':' . __LINE__ . ": ?????????????????????????? existing: $job->{id}\n";
+        my $notes = decode_json $job->{notes};
+        my $gru_id = $notes->{gru_id} or next;
+        my $job_args = decode_json $job->{args};
+        if ($job->{state} eq 'finished') {
+            # same task was run less than 1 minute ago and finished, nothing to do
+            @$jobs = ();
+            return;
+        }
+        $self->_add_jobs_to_gru_task($gru_id, $jobs);
+        last;
+    }
+}
+
+sub _add_jobs_to_gru_task ($self, $gru_id, $jobs) {
+    # warn __PACKAGE__ . ':' . __LINE__ . ": ================== _add_jobs_to_gru_task $gru_id $job->{created}\n";
+    my $schema = OpenQA::Schema->singleton;
+    while (@$jobs) {
+        my $id = $jobs->[0];
+        # Add job to existing gru task with the same args
+        # warn __PACKAGE__ . ':' . __LINE__ . ": !!! assigning job $id to gru $gru_id\n";
+        my $gru_dep = eval { $schema->resultset('GruDependencies')->create({job_id => $id, gru_task_id => $gru_id}); };
+        unless ($gru_dep) {
+            my $error = $@;
+            # warn __PACKAGE__ . ':' . __LINE__ . ": ???????????? ERROR: $error\n";
+            if ($error
+                =~ m/insert or update on table "gru_dependencies" violates foreign key constraint "gru_dependencies_fk_gru_task_id"/i
+              )
+            {
+                # if the GruTask was already deleted meanwhile, we can skip
+                # the rest of the jobs, since the wanted task was done
+                log_debug("GruTask $gru_id already gone, skip assigning jobs (message: $@)");
+                @$jobs = ();
+                last;
+            }
+            die $error;
+        }
+        shift @$jobs;
```

Comment:
```suggestion
            my $error = $@;
            die $error unless $error
                =~ m/insert or update on table "gru_dependencies" violates foreign key constraint "gru_dependencies_fk_gru_task_id"/i;
            # if the GruTask was already deleted meanwhile, we can skip
            # the rest of the jobs, since the wanted task was done
            log_debug("GruTask $gru_id already gone, skip assigning jobs (message: $@)");
            @$jobs = ();
            last;
        }
        shift @$jobs;
```